binutils/binutils/doc/strings.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "STRINGS 1"
.TH STRINGS 1 2023-07-30 binutils-2.41 "GNU Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
strings \- print the sequences of printable characters in files
.SH SYNOPSIS
.IX Header "SYNOPSIS"
strings [\fB\-afovV\fR] [\fB\-\fR\fImin-len\fR]
        [\fB\-n\fR \fImin-len\fR] [\fB\-\-bytes=\fR\fImin-len\fR]
        [\fB\-t\fR \fIradix\fR] [\fB\-\-radix=\fR\fIradix\fR]
        [\fB\-e\fR \fIencoding\fR] [\fB\-\-encoding=\fR\fIencoding\fR]
        [\fB\-U\fR \fImethod\fR] [\fB\-\-unicode=\fR\fImethod\fR]
        [\fB\-\fR] [\fB\-\-all\fR] [\fB\-\-print\-file\-name\fR]
        [\fB\-T\fR \fIbfdname\fR] [\fB\-\-target=\fR\fIbfdname\fR]
        [\fB\-w\fR] [\fB\-\-include\-all\-whitespace\fR]
        [\fB\-s\fR] [\fB\-\-output\-separator\fR \fIsep_string\fR]
        [\fB\-\-help\fR] [\fB\-\-version\fR] \fIfile\fR...
.SH DESCRIPTION
.IX Header "DESCRIPTION"
For each \fIfile\fR given, GNU \fBstrings\fR prints the
printable character sequences that are at least 4 characters long (or
the number given with the options below) and are followed by an
unprintable character.
.PP
Depending upon how the strings program was configured it will default
to either displaying all the printable sequences that it can find in
each file, or only those sequences that are in loadable, initialized
data sections.  If the file type is unrecognizable, or if strings is
reading from stdin then it will always display all of the printable
sequences that it can find.
.PP
For backwards compatibility any file that occurs after a command-line
option of just \fB\-\fR will also be scanned in full, regardless of
the presence of any \fB\-d\fR option.
.PP
\&\fBstrings\fR is mainly useful for determining the contents of
non-text files.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-a\fR 4
.IX Item "-a"
.PD 0
.IP \fB\-\-all\fR 4
.IX Item "--all"
.IP \fB\-\fR 4
.IX Item "-"
.PD
Scan the whole file, regardless of what sections it contains or
whether those sections are loaded or initialized.  Normally this is
the default behaviour, but strings can be configured so that the
\&\fB\-d\fR is the default instead.
.Sp
The \fB\-\fR option is position dependent and forces strings to
perform full scans of any file that is mentioned after the \fB\-\fR
on the command line, even if the \fB\-d\fR option has been
specified.
.IP \fB\-d\fR 4
.IX Item "-d"
.PD 0
.IP \fB\-\-data\fR 4
.IX Item "--data"
.PD
Only print strings from initialized, loaded data sections in the
file.  This may reduce the amount of garbage in the output, but it
also exposes the strings program to any security flaws that may be
present in the BFD library used to scan and load sections.  Strings
can be configured so that this option is the default behaviour.  In
such cases the \fB\-a\fR option can be used to avoid using the BFD
library and instead just print all of the strings found in the file.
.IP \fB\-f\fR 4
.IX Item "-f"
.PD 0
.IP \fB\-\-print\-file\-name\fR 4
.IX Item "--print-file-name"
.PD
Print the name of the file before each string.
.IP \fB\-\-help\fR 4
.IX Item "--help"
Print a summary of the program usage on the standard output and exit.
.IP \fB\-\fR\fImin-len\fR 4
.IX Item "-min-len"
.PD 0
.IP "\fB\-n\fR \fImin-len\fR" 4
.IX Item "-n min-len"
.IP \fB\-\-bytes=\fR\fImin-len\fR 4
.IX Item "--bytes=min-len"
.PD
Print sequences of displayable characters that are at least
\&\fImin-len\fR characters long.  If not specified a default minimum
length of 4 is used.  The distinction between displayable and
non-displayable characters depends upon the setting of the 
\&\fB\-e\fR and \fB\-U\fR options.  Sequences are always terminated
at control characters such as new-line and carriage-return, but not
the tab character.
.IP \fB\-o\fR 4
.IX Item "-o"
Like \fB\-t o\fR.  Some other versions of \fBstrings\fR have \fB\-o\fR
act like \fB\-t d\fR instead.  Since we can not be compatible with both
ways, we simply chose one.
.IP "\fB\-t\fR \fIradix\fR" 4
.IX Item "-t radix"
.PD 0
.IP \fB\-\-radix=\fR\fIradix\fR 4
.IX Item "--radix=radix"
.PD
Print the offset within the file before each string.  The single
character argument specifies the radix of the offset\-\-\-\fBo\fR for
octal, \fBx\fR for hexadecimal, or \fBd\fR for decimal.
.IP "\fB\-e\fR \fIencoding\fR" 4
.IX Item "-e encoding"
.PD 0
.IP \fB\-\-encoding=\fR\fIencoding\fR 4
.IX Item "--encoding=encoding"
.PD
Select the character encoding of the strings that are to be found.
Possible values for \fIencoding\fR are: \fBs\fR = single\-7\-bit\-byte
characters (default), \fBS\fR =
single\-8\-bit\-byte characters, \fBb\fR = 16\-bit bigendian, \fBl\fR =
16\-bit littleendian, \fBB\fR = 32\-bit bigendian, \fBL\fR = 32\-bit
littleendian.  Useful for finding wide character strings. (\fBl\fR
and \fBb\fR apply to, for example, Unicode UTF\-16/UCS\-2 encodings).
.IP "\fB\-U\fR \fI[d|i|l|e|x|h]\fR" 4
.IX Item "-U [d|i|l|e|x|h]"
.PD 0
.IP \fB\-\-unicode=\fR\fI[default|invalid|locale|escape|hex|highlight]\fR 4
.IX Item "--unicode=[default|invalid|locale|escape|hex|highlight]"
.PD
Controls the display of UTF\-8 encoded multibyte characters in strings.
The default (\fB\-\-unicode=default\fR) is to give them no special
treatment, and instead rely upon the setting of the
\&\fB\-\-encoding\fR option.  The other values for this option
automatically enable \fB\-\-encoding=S\fR.
.Sp
The \fB\-\-unicode=invalid\fR option treats them as non-graphic
characters and hence not part of a valid string.  All the remaining
options treat them as valid string characters.
.Sp
The \fB\-\-unicode=locale\fR option displays them in the current
locale, which may or may not support UTF\-8 encoding.  The
\&\fB\-\-unicode=hex\fR option displays them as hex byte sequences
enclosed between \fI<>\fR characters.  The \fB\-\-unicode=escape\fR
option displays them as escape sequences (\fI\euxxxx\fR) and the
\&\fB\-\-unicode=highlight\fR option displays them as escape sequences
highlighted in red (if supported by the output device).  The colouring
is intended to draw attention to the presence of unicode sequences
where they might not be expected.
.IP "\fB\-T\fR \fIbfdname\fR" 4
.IX Item "-T bfdname"
.PD 0
.IP \fB\-\-target=\fR\fIbfdname\fR 4
.IX Item "--target=bfdname"
.PD
Specify an object code format other than your system's default format.
.IP \fB\-v\fR 4
.IX Item "-v"
.PD 0
.IP \fB\-V\fR 4
.IX Item "-V"
.IP \fB\-\-version\fR 4
.IX Item "--version"
.PD
Print the program version number on the standard output and exit.
.IP \fB\-w\fR 4
.IX Item "-w"
.PD 0
.IP \fB\-\-include\-all\-whitespace\fR 4
.IX Item "--include-all-whitespace"
.PD
By default tab and space characters are included in the strings that
are displayed, but other whitespace characters, such a newlines and
carriage returns, are not.  The \fB\-w\fR option changes this so
that all whitespace characters are considered to be part of a string.
.IP \fB\-s\fR 4
.IX Item "-s"
.PD 0
.IP \fB\-\-output\-separator\fR 4
.IX Item "--output-separator"
.PD
By default, output strings are delimited by a new-line. This option
allows you to supply any string to be used as the output record
separator.  Useful with \-\-include\-all\-whitespace where strings
may contain new-lines internally.
.IP \fB@\fR\fIfile\fR 4
.IX Item "@file"
Read command-line options from \fIfile\fR.  The options read are
inserted in place of the original @\fIfile\fR option.  If \fIfile\fR
does not exist, or cannot be read, then the option will be treated
literally, and not removed.
.Sp
Options in \fIfile\fR are separated by whitespace.  A whitespace
character may be included in an option by surrounding the entire
option in either single or double quotes.  Any character (including a
backslash) may be included by prefixing the character to be included
with a backslash.  The \fIfile\fR may itself contain additional
@\fIfile\fR options; any such options will be processed recursively.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBar\fR\|(1), \fBnm\fR\|(1), \fBobjdump\fR\|(1), \fBranlib\fR\|(1), \fBreadelf\fR\|(1)
and the Info entries for \fIbinutils\fR.
.SH COPYRIGHT
.IX Header "COPYRIGHT"
Copyright (c) 1991\-2023 Free Software Foundation, Inc.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, with no Front-Cover Texts, and with no
Back-Cover Texts.  A copy of the license is included in the
section entitled "GNU Free Documentation License".


gcc/NEWS
This file contains information about GCC releases which has been generated
automatically from the online release notes.  It covers releases of GCC
(and the former EGCS project) since EGCS 1.0, on the line of development
that led to GCC 3. For information on GCC 2.8.1 and older releases of GCC 2,
see ONEWS.

======================================================================
http://gcc.gnu.org/gcc-13/index.html
                             GCC 13 Release Series

   July 27, 2023

   The GCC developers are pleased to announce the release of GCC 13.2.

   This release is a bug-fix release, containing fixes for regressions in
   GCC 13.1 relative to previous releases of GCC.

Release History

   GCC 13.2
          July 27, 2023 ([1]changes, [2]documentation)

   GCC 13.1
          April 26, 2023 ([3]changes, [4]documentation)

References and Acknowledgements

   GCC used to stand for the GNU C Compiler, but since the compiler
   supports several other languages aside from C, it now stands for the
   GNU Compiler Collection.

   A list of [5]successful builds is updated as new information becomes
   available.

   The GCC developers would like to thank the numerous people that have
   contributed new features, improvements, bug fixes, and other changes as
   well as test results to GCC. This [6]amazing group of volunteers is
   what makes GCC successful.

   For additional information about GCC please refer to the [7]GCC project
   web site or contact the [8]GCC development mailing list.

   To obtain GCC please use [9]our mirror sites or [10]our version control
   system.


    For questions related to the use of GCC, please consult these web
    pages and the [11]GCC manuals. If that fails, the
    [12]gcc-help@gcc.gnu.org mailing list might help. Comments on these
    web pages and the development of GCC are welcome on our developer
    list at [13]gcc@gcc.gnu.org. All of [14]our lists have public
    archives.

   Copyright (C) [15]Free Software Foundation, Inc. Verbatim copying and
   distribution of this entire article is permitted in any medium,
   provided this notice is preserved.

   These pages are [16]maintained by the GCC team. Last modified
   2023-07-27.

References

   1. http://gcc.gnu.org/gcc-13/changes.html
   2. http://gcc.gnu.org/onlinedocs/13.2.0/
   3. http://gcc.gnu.org/gcc-13/changes.html
   4. http://gcc.gnu.org/onlinedocs/13.1.0/
   5. http://gcc.gnu.org/gcc-13/buildstat.html
   6. http://gcc.gnu.org/onlinedocs/gcc-13.1.0/gcc/Contributors.html
   7. http://gcc.gnu.org/index.html
   8. mailto:gcc@gcc.gnu.org
   9. http://gcc.gnu.org/mirrors.html
  10. http://gcc.gnu.org/git.html
  11. https://gcc.gnu.org/onlinedocs/
  12. mailto:gcc-help@gcc.gnu.org
  13. mailto:gcc@gcc.gnu.org
  14. https://gcc.gnu.org/lists.html
  15. https://www.fsf.org/
  16. https://gcc.gnu.org/about.html
======================================================================
http://gcc.gnu.org/gcc-13/changes.html
                             GCC 13 Release Series
                        Changes, New Features, and Fixes

   This page is a "brief" summary of some of the huge number of
   improvements in GCC 13. You may also want to check out our [1]Porting
   to GCC 13 page and the [2]full GCC documentation.

Caveats

     * OpenMP offloading to Intel MIC has been removed.
     * The support for the cr16-elf, tilegx*-linux, tilepro*-linux,
       hppa[12]*-*-hpux10*, hppa[12]*-*-hpux11* and m32c-rtems
       configurations has been removed.
     * Support for Solaris 11.3 (*-*-solaris2.11.3) has been declared
       obsolete. The next release of GCC will have corresponding code
       permanently removed. Details can be found in the [3]announcement.
     * Support for emitting the STABS debugging format (including the
       -gstabs and -gxcoff options) has been removed. (This means the dbx
       debugger is no longer supported, either.)
     * Legacy debug info compression option -gz=zlib-gnu was removed and
       the option is ignored right now.
     * [4]-Warray-bounds=2 will no longer issue warnings for out of bounds
       accesses to trailing struct members of one-element array type
       anymore. Instead it diagnoses accesses to trailing arrays according
       to -fstrict-flex-arrays.
     * [5]-fanalyzer is still only suitable for analyzing C code. In
       particular, using it on C++ is unlikely to give meaningful output.
     * In the arm port, support for the iWMMXt extensions, enabled through
       -mcpu=iwmmxt, has been deprecated and will be removed in a future
       release. This includes support for the [6]iWMMXt built-in
       functions.
     * For C++, construction of the global iostream objects std::cout,
       std::cin, etc. is now done inside the standard library, instead of
       in every source file that includes the <iostream> header. This
       change improves the start-up performance of C++ programs, but it
       means that code compiled with GCC 13.1 will crash if the correct
       version of libstdc++.so is not used at runtime. See the
       [7]documentation about using the right libstdc++.so at runtime.
       Future GCC releases will mitigate the problem so that the program
       cannot be run at all with an older libstdc++.so.

General Improvements

     * [8]OpenMP
          + Reverse offload is now supported with AMD GCN and nvptx
            devices. Additionally, the requires handling has been improved
            and all clauses are now accepted. If a requirement cannot be
            fulfilled for an accessible device, this device is excluded
            from the list of available devices. This may imply that the
            only device left is the host (the initial device). In
            particular, unified_address and unified_shared_memory are
            unsupported by all non-host devices.
          + OpenMP 5.0: Fortran now supports some non-rectangular loop
            nests; for C/C++, the support was added in GCC 11.

cloog/osl/source/strings.c

    /*+-----------------------------------------------------------------**
     **                       OpenScop Library                          **
     **-----------------------------------------------------------------**
     **                           strings.c                             **
     **-----------------------------------------------------------------**
     **                   First version: 13/07/2011                     **
     **-----------------------------------------------------------------**

 
 *****************************************************************************
 * OpenScop: Structures and formats for polyhedral tools to talk together    *
 *****************************************************************************
 *    ,___,,_,__,,__,,__,,__,,_,__,,_,__,,__,,___,_,__,,_,__,                *
 *    /   / /  //  //  //  // /   / /  //  //   / /  // /  /|,_,             *
 *   /   / /  //  //  //  // /   / /  //  //   / /  // /  / / /\             *
 *  |~~~|~|~~~|~~~|~~~|~~~|~|~~~|~|~~~|~~~|~~~|~|~~~|~|~~~|/_/  \            *
 *  | G |C| P | = | L | P |=| = |C| = | = | = |=| = |=| C |\  \ /\           *
 *  | R |l| o | = | e | l |=| = |a| = | = | = |=| = |=| L | \# \ /\          *
 *  | A |a| l | = | t | u |=| = |n| = | = | = |=| = |=| o | |\# \  \         *
 *  | P |n| l | = | s | t |=| = |d| = | = | = | |   |=| o | | \# \  \        *
 *  | H | | y |   | e | o | | = |l|   |   | = | |   | | G | |  \  \  \       *
 *  | I | |   |   | e |   | |   | |   |   |   | |   | |   | |   \  \  \      *
 *  | T | |   |   |   |   | |   | |   |   |   | |   | |   | |    \  \  \     *
 *  | E | |   |   |   |   | |   | |   |   |   | |   | |   | |     \  \  \    *
 *  | * |*| * | * | * | * |*| * |*| * | * | * |*| * |*| * | /      \* \  \   *
 *  | O |p| e | n | S | c |o| p |-| L | i | b |r| a |r| y |/        \  \ /   *
 *  '---'-'---'---'---'---'-'---'-'---'---'---'-'---'-'---'          '--'    *
 *                                                                           *
 * Copyright (C) 2008 University Paris-Sud 11 and INRIA                      *
 *                                                                           *
 * (3-clause BSD license)                                                    *
 * Redistribution and use in source  and binary forms, with or without       *
 * modification, are permitted provided that the following conditions        *
 * are met:                                                                  *
 *                                                                           *
 * 1. Redistributions of source code must retain the above copyright notice, *
 *    this list of conditions and the following disclaimer.                  *
 * 2. Redistributions in binary form must reproduce the above copyright      *
 *    notice, this list of conditions and the following disclaimer in the    *
 *    documentation and/or other materials provided with the distribution.   *
 * 3. The name of the author may not be used to endorse or promote products  *
 *    derived from this software without specific prior written permission.  *
 *                                                                           *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR      *
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES *
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.   *
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,          *
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT  *
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, *
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY     *
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT       *
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF  *
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.         *
 *                                                                           *
 * OpenScop Library, a library to manipulate OpenScop formats and data       *
 * structures. Written by:                                                   *
 * Cedric Bastoul     <Cedric.Bastoul@u-psud.fr> and                         *
 * Louis-Noel Pouchet <Louis-Noel.pouchet@inria.fr>                          *
 *                                                                           *
 *****************************************************************************/

# include <stdlib.h>
# include <stdio.h>
# include <ctype.h>
# include <string.h>

# include <osl/macros.h>
# include <osl/util.h>
# include <osl/interface.h>
# include <osl/strings.h>


/*+***************************************************************************
 *                          Structure display function                       *
 *****************************************************************************/


/**
 * osl_strings_idump function:
 * this function displays an array of strings into a file (file, possibly
 * stdout) in a way that trends to be understandable. It includes an
 * indentation level (level) in order to work with others
 * idump functions.
 * \param[in] file    The file where the information has to be printed.
 * \param[in] strings The array of strings that has to be printed.
 * \param[in] level   Number of spaces before printing, for each line.
 */
void osl_strings_idump(FILE * file, osl_strings_p strings, int level) {
  int i, nb_strings;
  
  for (i = 0; i < level; i++)
    fprintf(file, "|\t");
  
  if (strings != NULL) {
    nb_strings = osl_strings_size(strings);
    fprintf(file, "+-- osl_strings_t:");
    for (i = 0; i < nb_strings; i++)
      fprintf(file, " %s", strings->string[i]);
    fprintf(file, "\n");
  }
  else
    fprintf(file, "+-- NULL strings\n");

  // A blank line.
  for (i = 0; i <= level; i++)
    fprintf(file, "|\t");
  fprintf(file, "\n");
}


/**
 * osl_strings_dump function:
 * this function prints the content of an osl_strings_t structure
 * (*strings) into a file (file, possibly stdout).
 * \param[in] file    The file where the information has to be printed.
 * \param[in] strings The strings structure which has to be printed.
 */
void osl_strings_dump(FILE * file, osl_strings_p strings) {
  osl_strings_idump(file, strings, 0);
}


/**
 * osl_strings_sprint function:
 * this function prints the content of an osl_strings_t structure
 * (*strings) into a string (returned) in the OpenScop textual format.
 * \param[in] strings The strings structure which has to be printed.
 * \return A string containing the OpenScop dump of the strings structure.
 */
char * osl_strings_sprint(osl_strings_p strings) {
  int i;
  int high_water_mark = OSL_MAX_STRING;
  char * string = NULL;
  char buffer[OSL_MAX_STRING];

  OSL_malloc(string, char *, high_water_mark * sizeof(char));
  string[0] = '\0';
   
  if (strings != NULL) {
    for (i = 0; i < osl_strings_size(strings); i++) {
      sprintf(buffer, "%s", strings->string[i]);
      osl_util_safe_strcat(&string, buffer, &high_water_mark);
      if (i < osl_strings_size(strings) - 1)
        osl_util_safe_strcat(&string, " ", &high_water_mark);
    }
    sprintf(buffer, "\n");
    osl_util_safe_strcat(&string, buffer, &high_water_mark);
  }
  else {
    sprintf(buffer, "# NULL strings\n");
    osl_util_safe_strcat(&string, buffer, &high_water_mark);
  }

  return string;
}


/**
 * osl_strings_print function:
 * this function prints the content of an osl_strings_t structure
 * (*body) into a file (file, possibly stdout) in the OpenScop format.
 * \param[in] file    File where informations are printed.
 * \param[in] strings The strings whose information has to be printed.
 */
void osl_strings_print(FILE * file, osl_strings_p strings) {
  char * string;
  
  string = osl_strings_sprint(strings);
  if (string != NULL) {
    fprintf(file, "%s", string);
    free(string);
  }
}


/*+***************************************************************************
 *                          Structure display function                       *
 *****************************************************************************/


/**
 * osl_strings_sread function:
 * this function reads a strings structure from a string complying to the
 * OpenScop textual format and returns a pointer to this strings structure.
 * The input string should only contain the list of strings this function
 * has to read (comments at the end of the line are accepted). The input
 * parameter is updated to the position in the input string this function
 * reach right after reading the strings structure.
 * \param[in,out] input The input string where to find a strings structure.
 *                      Updated to the position after what has been read.
 * \return A pointer to the strings structure that has been read.
 */
osl_strings_p osl_strings_sread(char ** input) {
  char tmp[OSL_MAX_STRING];
  char * s;
  char ** string = NULL;
  int nb_strings;
  int i, count;
  osl_strings_p strings = NULL;

  // Skip blank/commented lines and spaces before the strings.
  osl_util_sskip_blank_and_comments(input);
  
  // Count the actual number of strings.
  nb_strings = 0;
  s = *input;
  while (1) {
    for (count = 0; *s && !isspace(*s) && *s != '#'; count++)
      s++;
    
    if (count != 0)
      nb_strings++;

    if ((!*s) || (*s == '#') || (*s == '\n'))
      break;
    else
      s++;
  }

  if (nb_strings > 0) {
    // Allocate the array of strings. Make it NULL-terminated.
    OSL_malloc(string, char **, sizeof(char *) * (nb_strings + 1));
    string[nb_strings] = NULL;

    // Read the desired number of strings.
    s = *input;
    for (i = 0; i < nb_strings; i++) {
      for (count = 0; *s && !isspace(*s) && *s != '#'; count++)
	tmp[count] = *(s++);
      tmp[count] = '\0';
      OSL_strdup(string[i], tmp);
      if (*s != '#')
	s++;
    }

    // Update the input pointer to the end of the strings structure.
    *input = s;

    // Build the strings structure
    strings = osl_strings_malloc();
    strings->string = string;
  }

  return strings;
}


/**
 * osl_strings_read function.
 * this function reads a strings structure from a file (possibly stdin)
 * complying to the OpenScop textual format and returns a pointer to this
 * structure.
 * parameter nb_strings).
 * \param[in] file The file where to read the strings structure.
 * \return The strings structure that has been read.
 */
osl_strings_p osl_strings_read(FILE * file) {
  char buffer[OSL_MAX_STRING], * start;
  osl_strings_p strings;

  start = osl_util_skip_blank_and_comments(file, buffer);
  strings = osl_strings_sread(&start);

  return strings;
}
  

/*+***************************************************************************
 *                    Memory allocation/deallocation function                *
 *****************************************************************************/


/**
 * osl_strings_malloc function:
 * This function allocates the memory space for an osl_strings_t
 * structure and sets its fields with default values. Then it returns a
 * pointer to the allocated space.
 * \return A pointer to an empty strings structure with fields set to
 *         default values.
 */
osl_strings_p osl_strings_malloc() {
  osl_strings_p strings;

  OSL_malloc(strings, osl_strings_p, sizeof(osl_strings_t));
  strings->string = NULL;

  return strings;
}


/**
 * osl_strings_free function:
 * this function frees the allocated memory for a strings data structure.
 * \param[in] strings The strings structure we want to free.
 */
void osl_strings_free(osl_strings_p strings) {
  int i;

  if (strings != NULL) {
    if (strings->string != NULL) {
      i = 0;
      while (strings->string[i] != NULL) {
        free(strings->string[i]);
        i++;
      }
      free(strings->string);
    }
    free(strings);
  }
}


/*+***************************************************************************
 *                            Processing functions                           *
 *****************************************************************************/


/**
 * osl_strings_clone function.
 * this function builds and return a "hard copy" (not a pointer copy) of an
 * strings structure provided as parameter.
 * \param[in] strings The strings structure to clone.
 * \return The clone of the strings structure.
 */
osl_strings_p osl_strings_clone(osl_strings_p strings) {
  int i, nb_strings;
  osl_strings_p clone = NULL;
  
  if (strings == NULL)
    return NULL;

  clone = osl_strings_malloc();
  if ((nb_strings = osl_strings_size(strings)) == 0)
    return clone;

  OSL_malloc(clone->string, char **, (nb_strings + 1) * sizeof(char *));
  clone->string[nb_strings] = NULL;
  for (i = 0; i < nb_strings; i++)
    OSL_strdup(clone->string[i], strings->string[i]);

  return clone;
}


/**
 * osl_strings_equal function:
 * this function returns true if the two strings structures are the same
 * (content-wise), false otherwise.
 * \param[in] s1 The first strings structure.
 * \param[in] s2 The second strings structure.
 * \return 1 if s1 and s2 are the same (content-wise), 0 otherwise.
 */
int osl_strings_equal(osl_strings_p s1, osl_strings_p s2) {
  int i, nb_s1;
  
  if (s1 == s2)
    return 1;

  if (((s1 == NULL) && (s2 != NULL)) ||
      ((s1 != NULL) && (s2 == NULL)) ||
      ((nb_s1 = osl_strings_size(s1)) != osl_strings_size(s2)))
    return 0;

  for (i = 0; i < nb_s1; i++)
    if (strcmp(s1->string[i], s2->string[i]) != 0)
      return 0;

  return 1;
}


/**
 * osl_strings_size function:
 * this function returns the number of elements in the NULL-terminated
 * strings array of the strings structure.
 * \param[in] strings The strings structure we need to know the size.
 * \return The number of strings in the strings structure.
 */
int osl_strings_size(osl_strings_p strings) {
  int size = 0;

  if ((strings != NULL) && (strings->string != NULL)) {
    while (strings->string[size] != NULL) {
      size++;
    }
  }

  return size;
}


/**
 * osl_strings_encapsulate function:
 * this function builds a new strings structure to encapsulate the string
 * provided as a parameter (the reference to this string is used directly).
 * \param[in] string The string to encapsulate in a strings structure.
 * \return A new strings structure containing only the provided string.
 */
osl_strings_p osl_strings_encapsulate(char * string) {
  osl_strings_p capsule = osl_strings_malloc();
  
  OSL_malloc(capsule->string, char **, 2 * sizeof(char *));
  capsule->string[0] = string;
  capsule->string[1] = NULL;
  
  return capsule;
}


/**
 * osl_strings_interface function:
 * this function creates an interface structure corresponding to the strings
 * structure and returns it).
 * \return An interface structure for the strings structure.
 */
osl_interface_p osl_strings_interface() {
  osl_interface_p interface = osl_interface_malloc();
  
  interface->URI    = strdup(OSL_URI_STRINGS);
  interface->idump  = (osl_idump_f)osl_strings_idump;
  interface->sprint = (osl_sprint_f)osl_strings_sprint;
  interface->sread  = (osl_sread_f)osl_strings_sread;
  interface->malloc = (osl_malloc_f)osl_strings_malloc;
  interface->free   = (osl_free_f)osl_strings_free;
  interface->clone  = (osl_clone_f)osl_strings_clone;
  interface->equal  = (osl_equal_f)osl_strings_equal;

  return interface;
}


/**
 * osl_strings_generate function:
 * this function generates a new strings structure containing
 * 'nb_strings' strings of the form "prefixXX" where XX goes from 1 to
 * nb_strings.
 * \param[in] prefix     The prefix of the generated strings.
 * \param[in] nb_strings The number of strings to generate.
 * \return A new strings structure containing generated strings.
 */
osl_strings_p osl_strings_generate(char * prefix, int nb_strings) {
  char ** strings = NULL;
  char buff[strlen(prefix) + 16]; // TODO: better (log10(INT_MAX) ?) :-D.
  int i;
  osl_strings_p generated;

  if (nb_strings) {
    OSL_malloc(strings, char **, sizeof(char *) * (nb_strings + 1));
    strings[nb_strings] = NULL;
    for (i = 0; i < nb_strings; i++) {
      sprintf(buff, "%s%d", prefix, i + 1);
      strings[i] = strdup(buff);
      if (strings[i] == NULL)
        OSL_error("memory overflow");
    }
  }

  generated = osl_strings_malloc();
  generated->string = strings;
  return generated;
}
